crates/oxc_linter/src/rules/jsx_a11y/no_redundant_roles.rs

```diff
@@ -0,0 +1,135 @@
+use crate::{context::LintContext, rule::Rule, utils::has_jsx_prop_lowercase, AstNode};
+use oxc_ast::{
+    ast::{JSXAttributeItem, JSXAttributeValue, JSXElementName},
+    AstKind,
+};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::{self, Error},
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+use phf::{phf_map, phf_set};
+use std::collections::HashMap;
+
+#[derive(Debug, Error, Diagnostic)]
+#[error(
+    "eslint-plugin-jsx-a11y(no-redundant-roles): The element `{element}` has an implicit role of `{role}`. Defining this explicitly is redundant and should be avoided."
+)]
+#[diagnostic(
+    severity(warning),
+    help("Remove the redundant role `{role}` from the element `{element}`.")
+)]
+struct NoRedundantRolesDiagnostic {
+    #[label]
+    pub span: Span,
+    pub element: String,
+    pub role: String,
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct NoRedundantRoles;
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// Enforces that the explicit role property is not the same as implicit/default role property on element.
+    ///
+    /// ### Why is this bad?
+    /// Redundant roles can lead to confusion and verbosity in the codebase.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// // Bad
+    /// <nav role="navigation" />
+    ///
+    /// // Good
+    /// <nav />
+    /// ```
+    NoRedundantRoles,
+    correctness
+);
+
+static DEFAULT_ROLE_EXCEPTIONS: phf::Map<&'static str, phf::Set<&'static str>> = phf_map! {
+    "nav" => phf_set!{"navigation"},
+    "button" => phf_set!{"button"},
+    "body" => phf_set!{"document"},
+};
+
+static EMPTY_SET: phf::Set<&'static str> = phf_set! {};
+
+impl Rule for NoRedundantRoles {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let components: HashMap<String, String> = ctx
+            .settings()
+            .jsx_a11y
+            .components
+            .iter()
+            .map(|(k, v)| (k.to_string(), v.clone()))
```

Comment:
i can't seem to work out why we need this?

it seems expensive (it will run on every node), and aren't components already a hash map?

https://github.com/oxc-project/oxc/blob/856b9a59eb0dc60840e24da004dfbfe19c04027d/crates/oxc_linter/src/lib.rs#L62

i may be missing something here though

Reply:
Ahhhh, so true...
I don't know why I did this...

An improvement will be made to this part as well, as the `components` were also removed when fixed in 8e14a669c3d9f3eeb121f85d6da162f70d414b7d...

Reply:
thanks